### PR TITLE
feature: smtp driver for email backend

### LINKED ISSUE #4: Add SMTP Email Driver
Hi,
it would be interesting to have a smtp email backend driver for tyk dashboard for on-premise installation. We have no way to send email from tyk with our internal smtp.
I will pull a request for this feature.

Thanks

### PR DESCRIPTION
**Pull request by asoorm**
resolves: #4 

example usage
    
```
...
"email_backend": {
  "enable_email_notifications": true,
  "code": "smtp",
  "settings": {
    "SMTPUsername": "ahmet@tyk.io",
    "SMTPPassword": "abcde",
    "SMTPAddress": "smtp.gmail.com:587"
  },
},
...
```

### CODE REVIEW

File: init.go [hunk 1]
```diff
@@ -27,21 +28,21 @@ type EmailBackend interface {
...
 }
 
 func GetEmailBackend(code string) (EmailBackend, error) {
 	var thisInterface EmailBackend
 	var ok bool
 
-	thisInterface, ok = EmailBackendCodes[code]
-	if !ok {
-		return nil, errors.New("No backend with this code was found")
+	if thisInterface, ok = EmailBackendCodes[code]; !ok {
```
Comment (matiasinsaurralde) [inline init.go:46]: Can we use?
```go
if thisInterface, ok = EmailBackendCodes[code]; !ok {
		return nil, errors.New("no backend with this code was found")
}
return thisInterface, nil
```

File: smtp.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net"
+	"net/mail"
+	"net/smtp"
+)
+
+type SmtpEmailBackend struct {
```
Comment (matiasinsaurralde) [inline smtp.go]: `SMTPEmailBackend` to make the linter happy?

**Review by matiasinsaurralde — CHANGES_REQUESTED** [review]
Looks good, suggested some tweaks

**Comment by matiasinsaurralde** [comment]
The official [`SendMail` implementation](https://golang.org/src/net/smtp/smtp.go?s=9230:9308#L309) doesn't support `InsecureSkipVerify`, should we copy/tweak that code in order to allow self-signed certs when using this driver?
cc @martin @buger

**Comment by asoorm** [comment]
For me, `SendMail` is a nice wrapper - and hides away complexity. In order to make configurable, we need to do as @matiasinsaurralde suggested.

`&tls.Config{ServerName: c.serverName, InsecureSkipVerify: true}` or instead of hardcoding, get it from config.

For me, the MVP is to not support self-signed certs and go with current implementation (using stdlib as is). eventually - if it is a requested/required feature to be configurable, we could fork / copy the lib at a later date?

**Comment by lonelycode** [comment]
> For me, the MVP is to not support self-signed certs and go with current implementation (using stdlib as is). eventually - if it is a requested/required feature to be configurable, we could fork / copy the lib at a later date?

Agree, keep it simple, we can extend later

**Comment by asoorm** [comment]
This PR does not work where the smtp server requires SSL/TLS, but does not support STARTTLS.

Need to decide whether to implement & maintain a solution which elegantly supports `insecure`, `TLS/SSL` & `STARTTLS` - or whether to use a lib to do for us.

Any preferences @buger @dencoded @matiasinsaurralde ?

**Comment by buger** [comment]
Well, we for sure not should maintain our own lib, since Go smtp package already has TLS support https://golang.org/pkg/net/smtp/

Example usage: https://gist.github.com/jim3ma/b5c9edeac77ac92157f8f8affa290f45

So if TLS enabled in config, you just need to call StartTLS before Auth

**Comment by dencoded** [comment]
@asoorm  I think the simplest approach with using standard `smtp.SendMail` would be enough as this method does `StartTLS` and upgrades connection to TLS if server supports `STARTTLS` extension.

According to self-signed certs for smtp-server/insecure - I am not sure that it is a typical case.

**Comment by asoorm** [comment]
@dencoded In the case that the SMTP server requires TLS, but does not support `STARTTLS`, then will fail. It will not be able to upgrade the insecure connection.

**Comment by buger** [comment]
How common is not to support STARTLS and support TLS?

**Comment by asoorm** [comment]
@buger According to: https://www.fastmail.com/help/technical/ssltlsstarttls.html

> Currently, things seem relatively randomly split between people using SMTP SSL/TLS encrypted over port 465, and people using SMTP with STARTTLS upgrading over port 587.

So I know this is just one src, and I honestly don't know how reliable, but it looks like it could be 50/50 split - meaning we should either use a 3rd party lib, and if not, we will need to:

1. try `tls.Dial`
2. if fails, fallback to `net.Dial`
3. Copy rest of `smtp.SendMail`

**Comment by buger** [comment]
Lets start with STARTTLS only support, and will work based on user feedback. @Pinpin31 does it sounds good?

**Comment by Pinpin31** [comment]
Unfortunately, my smtp server don't support STARTTLS but require TLS. I try the pull request and the code fail.

There is  a comment on this github with a code that support STARTTLS or/and TLS : https://github.com/golang/go/issues/13694

I think the code works for all case. It works with my smtp server with parameters  :

  useTls     := true
  useStartTls:= false

```go
package main

import (
	"net/smtp"
	"fmt"
	"strconv"
	"net"
	"crypto/tls"
	"encoding/base64"
)

func main() {
  smtpHost   := "mail.{domain}.com"
  smtpPort   := 587
  smtpLogin  := "{user}@{domain}.com"
  smtpPasswd := "{password}"

  useTls     := true
  useStartTls:= false

  from := "{from@...}"
  to   := "{to@...}"
	title := "Just the sample message"

	body := "Sample message that is";

	header := make(map[string]string)
	header["From"] = from
	header["To"] = to
	header["Subject"] = title
	header["MIME-Version"] = "1.0"
	header["Content-Type"] = "text/plain; charset=\"utf-8\""
	header["Content-Transfer-Encoding"] = "base64"

	message := ""
	for k, v := range header {
		message += fmt.Sprintf("%s: %s\r\n", k, v)
	}
	message += "\r\n" + base64.StdEncoding.EncodeToString([]byte(body))

	conn, err := net.Dial("tcp", smtpHost+":"+strconv.Itoa(smtpPort))
	if err != nil {
		fmt.Println(err)
		return
	}

	// TLS
	tlsconfig := &tls.Config{
		InsecureSkipVerify: true,
		ServerName: smtpHost,
	}

	if useTls {
		conn = tls.Client(conn, tlsconfig)
	}

	client, err := smtp.NewClient(conn, smtpHost)
	if err != nil {
		fmt.Println(err)
		return
	}

	hasStartTLS, _ := client.Extension("STARTTLS")
	if useStartTls && hasStartTLS {
		fmt.Println("STARTTLS ...")
		if err = client.StartTLS(tlsconfig); err != nil {
			fmt.Println(err)
			return
		}
	}

	// Set up authentication information.
	auth := smtp.PlainAuth(
		"",
		smtpLogin,
		smtpPasswd,
		smtpHost,
	)

	if ok, _ := client.Extension("AUTH"); ok {
		if err := client.Auth(auth); err != nil {
			fmt.Printf("Error during AUTH %s\n", err)
			return
		}
	}
	fmt.Println("AUTH done")

	if err := client.Mail(from); err != nil {
		fmt.Printf("Error: %s\n", err)
		return
	}
	fmt.Println("FROM done")

	if err := client.Rcpt(to); err != nil {
		fmt.Printf("Error: %s\n", err)
		return
	}
	fmt.Println("TO done")

	w, err := client.Data()
	if err != nil {
		fmt.Printf("Error: %s\n", err)
		return
	}

	_, err = w.Write([]byte(message))
	if err != nil {
		fmt.Printf("Error: %s\n", err)
		return
	}

	err = w.Close()
	if err != nil {
		fmt.Printf("Error: %s\n", err)
		return
	}

	client.Quit()
}
```

**Comment by buger** [comment]
In general doing TLS connection if param is passed should not be an issue too, you just do not need to use SendEmail function, or probably just copy-paste its original implementation and add additionall attribute which should explicitly enable TSL https://golang.org/src/net/smtp/smtp.go?s=9230:9308#L319

**Comment by Pinpin31** [comment]
Maybe you could use this library : https://github.com/go-gomail/gomail
It seems possible to switch easily in TLS or not :

```go
package main

import (
    "gopkg.in/gomail.v2"
    "crypto/tls"
)

func main() {
    m := gomail.NewMessage()
    m.SetHeader("From", "nilton@domain.com")
    m.SetHeader("To", "nilton@domain2.com")
    m.SetHeader("Subject", "Hello!")
    m.SetBody("text/html", "Hello <b>Nilton OS</b>!")

    // Send the email to Bob
    d := gomail.NewPlainDialer("mail.domain.com", 587, "nilton@domain.com", "1234567")
    d.TLSConfig = &tls.Config{InsecureSkipVerify: true}
    if err := d.DialAndSend(m); err != nil {
        panic(err)
    }
}
```

**Comment by Pinpin31** [comment]
I do a commit with gomail in my first pull request if you want to test :
https://github.com/TykTechnologies/tyk-email-driver/pull/5

**Comment by asoorm** [comment]
I agree with going down the library route to avoid re-inventing the wheel, and adding unnecessary maintenance overhead. I will research `gomail` and some other libs today/tomorrow.

File: init.go [hunk 3]
```diff
@@ -27,20 +28,20 @@ type EmailBackend interface {
...
+		return nil, errors.New("no backend with this code was found")
```
Comment (dencoded) [inline init.go]: using predefined vars with errors could simplify unit-tests asserts

File: smtp.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	m.host = host
+	m.port, _ = strconv.Atoi(port)
+	m.user = user
+	m.pass = pass
+	m.isEnabled = true
+
+	return nil
+}
+
+func (m *SMTPEmailBackend) Send(emailMeta EmailMeta, emailData interface{}, textTemplateName TykTemplateName,
```
Comment (dencoded) [inline smtp.go]: we could cover this function by tests

File: smtp.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+type config struct {
+	host string
+	port int
+	user string
+	pass string
+}
+
+// Init receives the configs, validates them and sets on the SMTPEmailBackend struct for use by Send function
+func (m *SMTPEmailBackend) Init(conf map[string]string) error {
```
Comment (dencoded) [inline smtp.go:26]: we could cover this function by tests

File: smtp.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	type superEmailData struct {
+		Data   interface{}
+		Styles string
+	}
+
+	thisData := superEmailData{Data: emailData}
+	thisData.Styles = Styles
+
+	if err := PortalEmailTemplatesTXT.ExecuteTemplate(&txtDoc, string(textTemplateName), emailData); err != nil {
+		log.Error("error executing text template:", err)
```
Comment (dencoded) [inline smtp.go]: would be better to follow logging API `log.WithError(err).Error("error executing text template")`

File: smtp.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	msg := gomail.NewMessage()
+	msg.SetHeader("From", emailMeta.FromEmail, emailMeta.FromName)
+	msg.SetHeader("To", emailMeta.RecipientEmail)
+	msg.SetHeader("Subject", emailMeta.Subject)
+	msg.SetBody("text/plain", txtDoc.String())
+
+	dialer := gomail.NewDialer(m.host, m.port, m.user, m.pass)
+
+	if err := dialer.DialAndSend(msg); err != nil {
+		log.Error("error sending mail:", err)
```
Comment (dencoded) [inline smtp.go]: one more spot for `log.WithError(err).`

File: smtp.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+		return err
+	}
+
+	log.Debugf("email sent addr: %s, from: %s, to: %s", m.host, emailMeta.FromEmail, emailMeta.RecipientEmail)
```
Comment (dencoded) [inline smtp.go]: might be better to use `log.WithFields(...).Debug("email sent")` as logging library is supposed to simplify format output code like this

**Review by dencoded — CHANGES_REQUESTED** [review]
just add some simple unit test for initialization + some minor comments about logging

File: smtp.go (hunk 5 above)
Comment (asoorm) [reply smtp.go:26]: done. see `smtp_test.go`

**Review by dencoded — APPROVED** [review]

**Review by matiasinsaurralde — APPROVED** [review]
Looks good!

**Comment by Pinpin31** [comment]
I test this version with my smtp server and it works fine.
It's great !
Thanks

**Pull request merged**